File: q3.java
import java.util.*;

public class TestClass {
    // We use static variables to store grid dimensions, number of slabs, 
    // the grid data, total sum of elements, and the maximum covered sum found.
    static int N, K;
    static int[][] arr;
    static long totalSum = 0;
    static long maxCoveredSum = 0;
    static List<Slab> allSlabs = new ArrayList<>();

    // Slab class to represent a wooden slab covering two adjacent cells.
    static class Slab {
        int r1, c1, r2, c2; // Coordinates of the two cells covered
        long weight; // Sum of values of the two covered cells

        Slab(int r1, int c1, int r2, int c2, long weight) {
            this.r1 = r1; this.c1 = c1;
            this.r2 = r2; this.c2 = c2;
            this.weight = weight;
        }

        // Checks if this slab overlaps with another slab.
        boolean overlaps(Slab other) {
            // Overlap occurs if any cell of this slab is the same as any cell of the other slab.
            return (r1 == other.r1 && c1 == other.c1) || (r1 == other.r2 && c1 == other.c2) ||
                   (r2 == other.r1 && c2 == other.c1) || (r2 == other.r2 && c2 == other.c2);
        }
    }

    // Optimized backtracking function to find the maximum covered sum.
    // 'startIdx' is the index in 'allSlabs' from which to start searching.
    // 'kRem' is the remaining number of slabs to place.
    // 'currentSum' is the sum of weights of slabs placed in the current configuration.
    // 'usedIndices' stores indices of slabs already placed in the current configuration.
    static void backtrack(int startIdx, int kRem, long currentSum, Set<Integer> usedIndices) {
        if (kRem == 0) {
            maxCoveredSum = Math.max(maxCoveredSum, currentSum);
            return;
        }

        for (int i = startIdx; i < allSlabs.size(); i++) {
            Slab currentSlab = allSlabs.get(i);
            
            // Check for overlap with previously placed slabs.
            boolean overlaps = false;
            for (int index : usedIndices) {
                if (currentSlab.overlaps(allSlabs.get(index))) {
                    overlaps = true;
                    break;
                }
            }

            if (!overlaps) {
                // If no overlap, include the slab and recurse.
                usedIndices.add(i);
                backtrack(i + 1, kRem - 1, currentSum + currentSlab.weight, usedIndices);
                usedIndices.remove(i); // Backtrack and remove the slab for the next iteration.
            }
        }
    }

    public static void sumOfVisibleCells(int nVal, int kVal, int[][] arrVal) {
        N = nVal;
        K = kVal;
        arr = arrVal;
        
        // 1. Calculate the total sum of all elements in the grid.
        for (int[] row : arr) {
            for (int val : row) {
                totalSum += val;
            }
        }
        
        // 2. Generate all possible horizontal and vertical slabs.
        
        // Horizontal slabs
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N - 1; j++) {
                allSlabs.add(new Slab(i, j, i, j + 1, (long) arr[i][j] + arr[i][j + 1]));
            }
        }
        
        // Vertical slabs
        for (int i = 0; i < N - 1; i++) {
            for (int j = 0; j < N; j++) {
                allSlabs.add(new Slab(i, j, i + 1, j, (long) arr[i][j] + arr[i + 1][j]));
            }
        }
        
        // 3. Use backtracking to find the maximum covered sum with K non-overlapping slabs.
        // Since K is very small (<=8), this backtracking approach is feasible.
        backtrack(0, K, 0, new HashSet<>());
        
        // 4. The minimum visible sum is the total sum minus the maximum covered sum.
        System.out.println(totalSum - maxCoveredSum);
    }
    
    public static void main(String[] args) {
        // Main function handles input reading and calls sumOfVisibleCells.
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int k = sc.nextInt();
        int[][] arr = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                arr[i][j] = sc.nextInt();
            }
        }
        
        sumOfVisibleCells(n, k, arr);
        sc.close();
    }
}

File: q3.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

// Constants defined in the template
const int MAXN = 1000;
const int MAXK = 8; 

// Define a struct for Slab
struct Slab {
    int r1, c1, r2, c2;
    long long weight;

    // Helper to check for overlap.
    bool overlaps(const Slab& other) const {
        // A slab overlaps if they share any cell.
        return (r1 == other.r1 && c1 == other.c1) || (r1 == other.r2 && c1 == other.c2) ||
               (r2 == other.r1 && c2 == other.c1) || (r2 == other.r2 && c2 == other.c2);
    }
};

// Global variables to store all generated slabs, total sum, and maximum covered sum.
vector<Slab> allSlabs;
long long totalSum = 0;
long long maxCoveredSum = 0;
int N_global, K_global;

// Backtracking function to find the maximum weight of K non-overlapping slabs.
// We iterate through all slabs and recursively select K of them, ensuring no overlap.
void backtrack(int startIdx, int kRem, long long currentSum, vector<int>& usedIndices) {
    // Base case: K slabs have been placed.
    if (kRem == 0) {
        maxCoveredSum = max(maxCoveredSum, currentSum);
        return;
    }

    // Iterate through remaining slabs starting from startIdx.
    for (int i = startIdx; i < allSlabs.size(); ++i) {
        const Slab& currentSlab = allSlabs[i];
        
        // Check for overlap with previously placed slabs.
        bool overlaps = false;
        for (int idx : usedIndices) {
            if (currentSlab.overlaps(allSlabs[idx])) {
                overlaps = true;
                break;
            }
        }

        if (!overlaps) {
            // If no overlap, include the slab and recurse.
            usedIndices.push_back(i);
            // Recurse for the next slab, reducing kRem by 1 and increasing currentSum.
            backtrack(i + 1, kRem - 1, currentSum + currentSlab.weight, usedIndices);
            usedIndices.pop_back(); // Backtrack
        }
    }
}

// Main solving function.
void sumOfVisibleCells(int N, int K, int arr[][MAXN]) {
    N_global = N;
    K_global = K;
    
    // 1. Calculate total sum of all elements in the grid.
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            totalSum += arr[i][j];
        }
    }
    
    // 2. Generate all possible slabs.
    
    // Horizontal slabs (rows, col j to j+1)
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N - 1; ++j) {
            allSlabs.push_back({i, j, i, j + 1, (long long)arr[i][j] + arr[i][j+1]});
        }
    }
    
    // Vertical slabs (row i to i+1, cols)
    for (int i = 0; i < N - 1; ++i) {
        for (int j = 0; j < N; ++j) {
            allSlabs.push_back({i, j, i + 1, j, (long long)arr[i][j] + arr[i+1][j]});
        }
    }
    
    // 3. Run backtracking to find the maximum covered sum.
    vector<int> usedIndices;
    backtrack(0, K, 0, usedIndices);
    
    // 4. Output the minimum visible sum (Total Sum - Maximum Covered Sum).
    cout << totalSum - maxCoveredSum << endl;
}

// The main function from the template
int main() {
    // Fast IO
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, k;
    cin >> n >> k;

    // Input array
    int arr[MAXN][MAXN];

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> arr[i][j];
        }
    }

    sumOfVisibleCells(n, k, arr);

    return 0;
}
